core/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import com.google.common.collect.UnmodifiableIterator;
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.*;
+
+/**
+ * If this is an alias it groups all {@link IndexMetaData} instances sharing the same alias name together, otherwise
+ * if this is a concrete index it just encapsulates the {@link IndexMetaData} of that index.
+ */
+public class AliasOrIndex {
+
+    private final IndexMetaData concreteIndex;
+
+    private final String aliasName;
+    private final List<IndexMetaData> referenceIndexMetaDatas;
+
+    public AliasOrIndex(IndexMetaData indexMetaData) {
+        this.aliasName = null;
+        this.referenceIndexMetaDatas = null;
```

Comment:
no, because the calling side checks `#isAlias()` and for an index that will evaluate `false` and the caller side won 't invoke methods that interact with `this.referenceIndexMetaDatas `.

What I actually have been thinking about is to make `AliasOrIndex` an interface and have two implementations: `Index` and `Alias`, this way we don't need these confusing null checks and we can make the contract of this class simpler.